File: ios/Runner/EventChannelDemo.swift
//
//  EventChannelDemo.swift
//  Runner
//
//  Created by a on 2020/12/28.
//

import Foundation
import Flutter
import UIKit

public class EventChannelDemo: NSObject, FlutterStreamHandler {
    var channel: FlutterEventChannel?
    var count = 0
    var events: FlutterEventSink?
    var timer: Timer?
    
    public override init() {
        super.init()
    }
    
    convenience init(messenger: FlutterBinaryMessenger) {
        self.init()
        
        channel = FlutterEventChannel(name: "com.mufeng.flutterNativeChannel.EventChannel", binaryMessenger: messenger)
        channel?.setStreamHandler(self)
        startTimer()
    }
    
    func startTimer() {
        timer = Timer.scheduledTimer(timeInterval: 1, target: self,selector: #selector(self.tickDown),userInfo: nil,repeats: true)
    }
    
    @objc func tickDown(){
        count += 1
        let args = ["count": count]
        if(events != nil){
            events!(args)
        }
        if(count >= 30){
            if(events != nil){
                events!(FlutterError.init(code: "400", message: "超过30秒,倒计时结束", details: nil))
            }
            timer?.fireDate = Date.distantFuture
            timer?.invalidate()
        }
    }
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.events = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.events = nil
        timer?.invalidate()
        self.timer = nil
        return nil
    }
}

File: android/app/src/main/kotlin/com/mufeng/flutter_native_channel/EventChannelDemo.kt
package com.mufeng.flutter_native_channel

import android.app.Activity
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import java.util.*
import kotlin.concurrent.timerTask

class EventChannelDemo(private var activity: Activity, messenger: BinaryMessenger) : EventChannel.StreamHandler {

    private var channel: EventChannel = EventChannel(messenger, "com.mufeng.flutterNativeChannel.EventChannel")
    private var count = 0
    private var events: EventChannel.EventSink? = null
    init {
        channel.setStreamHandler(this)
        startTimer()
    }

    private fun startTimer() {
        var timer = Timer().schedule(timerTask {
            count++
            val map = mapOf("count" to count)
            activity.runOnUiThread {
                events?.success(map)
            }
            if(count >= 30){
                activity.runOnUiThread {
                    events?.error("400", "超过30秒,倒计时结束", null)
                }
                cancel()
            }
        }, 0, 1000)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        this.events = events
    }

    override fun onCancel(arguments: Any?) {
        this.events = null
    }
}